caproto/server/autosave.py

```diff
@@ -0,0 +1,175 @@
+        return {
+            pvname: {'value': pvprop.value,
+                     'fields': dict(get_autosave_fields(pvprop))
+                     }
+            for pvname, pvprop in self.find_autosave_properties()
+        }
+
+    async def restore_from_file(self, filename):
+        """Restore from the autosave file."""
+        filename = pathlib.Path(filename)
+        if not filename.exists():
+            self.log.warning('Autosave file does not exist: %s', filename)
+            return
+
+        try:
+            with open(filename, 'rt') as f:
+                data = json.load(f)
+        except Exception:
+            self.log.exception('Failed to load JSON from %s', filename)
+            return
+
+        try:
+            await self.restore_values(data)
+        except Exception:
+            self.log.exception('Failed to restore values from %s', filename)
+            return
+
+    async def restore_values(self, data):
+        """Restore given the autosave file ``data`` dictionary."""
+        for pvname, info in data.items():
+            try:
+                pvprop = self.parent.pvdb[pvname]
+            except KeyError:
+                self.log.error('Autosave pvname not in database: %s', pvname)
+                continue
+
+            try:
+                await pvprop.write(info['value'])
+            except Exception as ex:
+                self.log.exception('Autosave restore failed: %s %s', pvname,
+                                   ex)
+            else:
+                self.log.info('Restored %s => %s', pvname, pvprop.value)
+
+            fields = info.get('fields', None) or {}
+            for field_name, field_value in fields.items():
+                try:
+                    field = getattr(pvprop.field_inst, field_name)
+                    await field.write(field_value)
+                except Exception as ex:
+                    self.log.exception(
+                        'Autosave restore failed: %s field %s = %s %s', pvname,
+                        field_name, field_value, ex)
+                else:
+                    self.log.info('Restored %s field %s => %s', pvname,
+                                  field_name, field.value)
+
+    async def save(self, stream, data=None):
+        """Save autosave ``data`` dictionary to the stream."""
+        if data is None:
+            data = self.prepare_data()
+
+        json.dump(data, stream)
+
+
+class AutosavedPVSpec(PVSpec):
+    """A hack to enhance the inflexible PVSpec with autosave settings."""
+    autosave = None
+
+    def _replace(self, **kwargs):
+        # NOTE: a further hack: this is a feature from `namedtuple` that's used
+        # to replace item(s) in the tuple, returning a newly modified copy.
+        # Here, we attach on `autosave` settings after performing the replace.
+        replaced = super()._replace(**kwargs)
+        replaced.autosave = self.autosave
+        return replaced
+
+
+def autosaved(pvprop, fields=None):
```

Comment:
Is there a way to make this fail loudly and helpfully if there is no `autosave_hook` running?

Reply:
Hmm, maybe. Good idea - I think this could be an issue for later.